Fixes #1223. Throttle force GC

eed3si9n opened the pull request:
This is take 2 of the force GC. This will run GC on a set interval or before cross building.

@cunei @jsuereth Do you have some test case that I can use to see if this fixes #1223 and that it won't blow up with `~compile`?

eed3si9n committed:
Fixes #1223. Uses JMX to find out when to force GC.

jsuereth commented on main/src/main/scala/sbt/EvaluateTask.scala:
```diff
@@ -322,12 +346,20 @@ object EvaluateTask {
+            if (opfc > config.maxObjectPendingFinalization) {
+              log.info(s"Forcing garbage collection... Objects pending finalization: $opfc")
+              // Force the detection of finalizers for scala.reflect weakhashsets
+              System.gc()
```
You may need to run this before the check willbe true...

eed3si9n commented:
Not sure why I close this, but the issue #1223 is still alive and kicking.
Maybe I need some throttling as mentioned in #1329.

jsuereth commented:
besides my one minor comment, I think this LGTM now.

jsuereth commented:
Also, please rebase.

eed3si9n committed:
Use interval throttling

JMX doesn't seem to work in reporting

    memoryMxBean.getObjectPendingFinalizationCount

At least for the test build that I used it always reports zero.

eed3si9n committed:
Force GC on cross building

eed3si9n commented:
/review @cunei @jsuereth

dwijnand commented:
Why not use `scala.concurrent.duration.Duration`/`FiniteDuration` instead of `Int`, which needs to be documented as "(in seconds)" everywhere?

Among other things it would also match your notes: `(default: 60s)`.

jsuereth commented on main/src/main/scala/sbt/GCUtil.scala:
```diff
@@ -0,0 +1,36 @@
+  def forceGcWithInterval(minForcegcInterval: Int, log: Logger): Unit =
+    {
+      val now = System.currentTimeMillis
+      val last = lastGcCheck.get
```
lazySet for performance, YO

jsuereth commented on main/src/main/scala/sbt/GCUtil.scala:
```diff
@@ -0,0 +1,36 @@
+      val last = lastGcCheck.get
+      // This throttles System.gc calls to interval
+      if (now - last > minForcegcInterval * 1000) {
+        lastGcCheck.set(now)
```
lazySet for performance, YO

eed3si9n committed:
use Duration to track minForcegcInterval

dwijnand commented:
Nice! Thank you.

eed3si9n commented:
I'm seeing

```
[error] (sbtRoot/*:scripted) Failed tests:
[error]     source-dependencies / macro-arg-dep-2-11
[error]     source-dependencies / macro-arg-dep-nested-2-11
[error] Total time: 723 s, completed May 6, 2015 1:30:32 AM
```

on Travis, but I can't reproduce it locally, so I'm going to merge this.
